File: backend/src/server/routes/catalog.py
from typing import List
import json
from pathlib import Path
from fastapi import APIRouter, HTTPException, Query
from ...models.catalog import (
    TechListResponse,
    WaferMapListResponse,
    ProcessOptionsResponse,
    ProcessContextResponse,
    ToolOptionsResponse,
    ToolProfileResponse,
)
from ...models.enums import Mode

router = APIRouter()

# Load strategies catalog
_STRATEGIES_CATALOG_PATH = Path(__file__).parent.parent.parent / "data" / "catalog" / "strategies.json"

def get_enabled_strategies() -> List[str]:
    """
    Load and return list of enabled strategy IDs from catalog.

    Returns:
        List of enabled strategy_id strings
    """
    try:
        with open(_STRATEGIES_CATALOG_PATH, 'r') as f:
            catalog = json.load(f)
            return [s["strategy_id"] for s in catalog["strategies"] if s.get("enabled", False)]
    except FileNotFoundError:
        # Fallback to default if catalog file not found
        return ["CENTER_EDGE"]
    except Exception as e:
        # Fallback to default if malformed
        return ["CENTER_EDGE"]

@router.get("/techs", response_model=TechListResponse)
async def list_techs():
    # Static placeholder data for v0
    return TechListResponse(techs=["28nm", "14nm", "7nm"])

@router.get("/wafer-maps", response_model=WaferMapListResponse)
async def list_wafer_maps(tech: str = Query(...)):
    # Placeholder static data aligned with OpenAPI
    return WaferMapListResponse(wafer_maps=[
        {"wafer_map_id": f"{tech}_standard", "tech": tech, "description": f"Standard {tech} wafer map"}
    ])

@router.get("/process-options", response_model=ProcessOptionsResponse)
async def list_process_options(tech: str = Query(...)):
    return ProcessOptionsResponse(process_options=[
        {
            "process_step": "LITHO",
            "intents": ["UNIFORMITY", "THICKNESS"],
            "modes": ["INLINE", "OFFLINE"]
        },
        {
            "process_step": "ETCH",
            "intents": ["CD_CONTROL", "PROFILE"],
            "modes": ["INLINE", "MONITOR"]
        }
    ])

@router.get("/process-context", response_model=ProcessContextResponse)
async def get_process_context(
    tech: str = Query(...),
    step: str = Query(...),
    intent: str = Query(...),
    mode: Mode = Query(...)
):
    # Derive criticality and constraints based on inputs
    criticality = "HIGH" if step == "LITHO" else "MEDIUM"
    min_points = 5 if criticality == "HIGH" else 3
    max_points = 25 if mode == "INLINE" else 50

    # Get enabled strategies from catalog
    enabled_strategies = get_enabled_strategies()

    return ProcessContextResponse(process_context={
        "process_step": step,
        "measurement_intent": intent,
        "mode": mode,
        "criticality": criticality,
        "min_sampling_points": min_points,
        "max_sampling_points": max_points,
        "allowed_strategy_set": enabled_strategies,
        "version": "1.0"
    })

@router.get("/tool-options", response_model=ToolOptionsResponse)
async def list_tool_options(
    tech: str = Query(...),
    step: str = Query(...),
    intent: str = Query(...)
):
    return ToolOptionsResponse(tool_options=[
        {"tool_type": "OPTICAL_METROLOGY", "vendor": "ASML", "model": "YieldStar"},
        {"tool_type": "SEM", "vendor": "AMAT", "model": "eSEM"}
    ])

@router.get("/tool-profile", response_model=ToolProfileResponse)
async def get_tool_profile(
    toolType: str = Query(...),
    vendor: str = Query(None),
    model: str = Query(None)
):
    # Tool capability based on type
    if toolType == "OPTICAL_METROLOGY":
        max_points = 49
        edge_supported = True
    else:  # SEM
        max_points = 25
        edge_supported = False

    return ToolProfileResponse(tool_profile={
        "tool_type": toolType,
        "vendor": vendor or "DEFAULT",
        "model": model,
        "coordinate_system_supported": ["DIE_GRID", "MM"],
        "max_points_per_wafer": max_points,
        "edge_die_supported": edge_supported,
        "ordering_required": False,
        "recipe_format": {"type": "JSON", "version": "1.0"},
        "forbidden_regions": [],
        "version": "1.0"
    })

File: frontend/src/api/catalog.ts
// frontend/src/api/catalog.ts

const API_BASE_URL = 'http://localhost:8080/v1';

// --- Types based on openapi.yaml ---
export interface TechListResponse {
  techs: string[];
}

export type Mode = 'INLINE' | 'OFFLINE' | 'MONITOR';

export interface ProcessOption {
  process_step: string;
  intents: string[];
  modes: Mode[];
}

export interface ProcessOptionsResponse {
  process_options: ProcessOption[];
}

export interface DiePoint {
  die_x: number;
  die_y: number;
}

export interface WaferMapSpec {
  wafer_size_mm: number;
  die_pitch_x_mm: number;
  die_pitch_y_mm: number;
  origin: 'CENTER';
  notch_orientation_deg: number;
  coordinate_system: 'DIE_GRID' | 'MM' | 'SHOT';
  valid_die_mask: {
    type: 'EDGE_EXCLUSION' | 'EXPLICIT_LIST';
    radius_mm?: number;
    valid_die_list?: DiePoint[];
  };
  version: string;
}

export interface WaferMapSummary {
  wafer_map_id: string;
  tech: string;
  description: string;
}

export interface WaferMapListResponse {
  wafer_maps: WaferMapSummary[];
}

export interface ProcessContext {
  process_step: string;
  measurement_intent: string;
  mode: Mode;
  criticality: 'HIGH' | 'MEDIUM' | 'LOW';
  min_sampling_points: number;
  max_sampling_points: number;
  allowed_strategy_set: string[];
  version: string;
}

export interface ProcessContextResponse {
  process_context: ProcessContext;
}

export type CoordSystemSupported = 'DIE_GRID' | 'MM' | 'SHOT';

export interface ToolProfile {
  tool_type: string;
  vendor: string;
  model?: string;
  coordinate_system_supported: CoordSystemSupported[];
  max_points_per_wafer: number;
  edge_die_supported: boolean;
  ordering_required: boolean;
  recipe_format: {
    type: 'JSON' | 'CSV' | 'TEXT';
    version: string;
  };
  version: string;
}

export interface ToolProfileResponse {
  tool_profile: ToolProfile;
}

export interface ToolOption {
  tool_type: string;
  vendor: string;
  model?: string;
}

export interface ToolOptionsResponse {
  tool_options: ToolOption[];
}

export interface Warning {
  code: string;
  message: string;
}

export interface SamplingOutput {
  sampling_strategy_id: string;
  selected_points: DiePoint[];
  point_tags?: string[];
  trace: {
    strategy_version: string;
    generated_at: string;
  };
}

export interface SamplingPreviewRequest {
  wafer_map_spec: WaferMapSpec;
  process_context: ProcessContext;
  tool_profile: ToolProfile;
  strategy: {
    strategy_id: string;
    strategy_config?: {
      common?: {
        edge_exclusion_mm?: number;
        target_point_count?: number;
        rotation_seed?: number;
      };
      advanced?: Record<string, any>;
    };
  };
}

export interface SamplingPreviewResponse {
  sampling_output: SamplingOutput;
  warnings: Warning[];
}

export interface SamplingScoreReport {
  coverage_score: number;
  statistical_score: number;
  risk_alignment_score: number;
  overall_score: number;
  warnings: Warning[];
  version: string;
}

export interface SamplingScoreRequest {
  wafer_map_spec: WaferMapSpec;
  process_context: ProcessContext;
  tool_profile: ToolProfile;
  sampling_output: SamplingOutput;
}

export interface SamplingScoreResponse {
  score_report: SamplingScoreReport;
}

export interface ToolRecipe {
  recipe_id: string;
  tool_type: string;
  recipe_payload: Record<string, any>;
  translation_notes?: string[];
  recipe_format_version: string;
}

export interface GenerateRecipeRequest {
  wafer_map_spec: WaferMapSpec;
  tool_profile: ToolProfile;
  sampling_output: SamplingOutput;
  score_report?: SamplingScoreReport; // Optional based on openapi.yaml
}

export interface GenerateRecipeResponse {
  tool_recipe: ToolRecipe;
  warnings: Warning[];
}

// --- API Client Functions ---

// Development mock data
const MOCK_DATA = {
  techs: ['28nm', '22nm', '16nm', '14nm', '10nm', '7nm', '5nm', '3nm'],
  waferMaps: [
    { wafer_map_id: 'standard_300mm', tech: '28nm', description: 'Standard 300mm wafer layout' },
    { wafer_map_id: 'high_density_300mm', tech: '28nm', description: 'High density 300mm layout' },
    { wafer_map_id: 'test_200mm', tech: '28nm', description: 'Test 200mm wafer layout' },
  ],
  processOptions: [
    { process_step: 'lithography', intents: ['critical_dimension', 'overlay'], modes: ['INLINE', 'OFFLINE'] as Mode[] },
    { process_step: 'etch', intents: ['depth', 'uniformity'], modes: ['INLINE', 'MONITOR'] as Mode[] },
    { process_step: 'metal', intents: ['resistance', 'thickness'], modes: ['OFFLINE', 'MONITOR'] as Mode[] },
  ],
  toolOptions: [
    { tool_type: 'SEM', vendor: 'ASML', model: 'eSEM-1000' },
    { tool_type: 'AFM', vendor: 'Bruker', model: 'Dimension-FastScan' },
    { tool_type: 'Optical', vendor: 'KLA', model: 'Archer-750' },
  ]
};

async function fetchAPI<T>(url: string, options?: RequestInit): Promise<T> {
  try {
    const response = await fetch(url, options);
    if (!response.ok) {
      const errorBody = await response.json().catch(() => ({ message: 'Unknown error' }));
      throw new Error(
        `HTTP error! status: ${response.status} - ${errorBody.error?.message || 'No error message'}`,
      );
    }
    return response.json();
  } catch (error) {
    console.warn(`API call failed: ${url}. Using mock data for development.`);
    throw error; // Let individual functions handle with mock data
  }
}

export const listTechs = async (): Promise<TechListResponse> => {
  try {
    return await fetchAPI<TechListResponse>(`${API_BASE_URL}/catalog/techs`);
  } catch {
    console.log('Using mock techs data');
    return Promise.resolve({ techs: MOCK_DATA.techs });
  }
};

export const listProcessOptions = async (tech: string): Promise<ProcessOptionsResponse> => {
  try {
    return await fetchAPI<ProcessOptionsResponse>(`${API_BASE_URL}/catalog/process-options?tech=${tech}`);
  } catch {
    console.log('Using mock process options data');
    return Promise.resolve({ process_options: MOCK_DATA.processOptions });
  }
};

export const listWaferMaps = async (tech: string): Promise<WaferMapListResponse> => {
  try {
    return await fetchAPI<WaferMapListResponse>(`${API_BASE_URL}/catalog/wafer-maps?tech=${tech}`);
  } catch {
    console.log('Using mock wafer maps data');
    return Promise.resolve({ wafer_maps: MOCK_DATA.waferMaps });
  }
};

// v1 Proposal: This endpoint is missing. We need a way to get the full WaferMapSpec from a wafer_map_id.
// For now, returning a mock spec.
export const getWaferMapSpec = async (waferMapId: string): Promise<WaferMapSpec> => {
  console.warn(`[MOCK] getWaferMapSpec called for ${waferMapId}. Returning mock data.`);
  return Promise.resolve({
    wafer_size_mm: 300,
    die_pitch_x_mm: 10,
    die_pitch_y_mm: 10,
    origin: 'CENTER',
    notch_orientation_deg: 270,
    coordinate_system: 'DIE_GRID',
    valid_die_mask: { type: 'EDGE_EXCLUSION', radius_mm: 145 },
    version: 'mock-v1',
  });
};

export const getProcessContext = async (params: { tech: string; step: string; intent: string; mode: Mode }): Promise<ProcessContextResponse> => {
  try {
    return await fetchAPI<ProcessContextResponse>(
      `${API_BASE_URL}/catalog/process-context?tech=${params.tech}&step=${params.step}&intent=${params.intent}&mode=${params.mode}`,
    );
  } catch {
    console.log('Using mock process context data');
    return Promise.resolve({
      process_context: {
        process_step: params.step,
        measurement_intent: params.intent,
        mode: params.mode,
        criticality: 'MEDIUM' as const,
        min_sampling_points: 10,
        max_sampling_points: 100,
        allowed_strategy_set: ['CENTER_EDGE', 'RANDOM_UNIFORM', 'GRID'],
        version: 'mock-v1'
      }
    });
  }
};

export const listToolOptions = async (params: { tech: string; step: string; intent: string }): Promise<ToolOptionsResponse> => {
  try {
    return await fetchAPI<ToolOptionsResponse>(
      `${API_BASE_URL}/catalog/tool-options?tech=${params.tech}&step=${params.step}&intent=${params.intent}`,
    );
  } catch {
    console.log('Using mock tool options data');
    return Promise.resolve({ tool_options: MOCK_DATA.toolOptions });
  }
};

export const getToolProfile = async (toolType: string): Promise<ToolProfileResponse> => {
  try {
    return await fetchAPI<ToolProfileResponse>(`${API_BASE_URL}/catalog/tool-profile?toolType=${toolType}`);
  } catch {
    console.log('Using mock tool profile data');
    return Promise.resolve({
      tool_profile: {
        tool_type: toolType,
        vendor: 'MockVendor',
        model: 'MockModel-1000',
        coordinate_system_supported: ['DIE_GRID', 'MM'] as CoordSystemSupported[],
        max_points_per_wafer: 1000,
        edge_die_supported: true,
        ordering_required: false,
        recipe_format: {
          type: 'JSON' as const,
          version: '1.0'
        },
        version: 'mock-v1'
      }
    });
  }
};

export const previewSampling = (requestBody: SamplingPreviewRequest) =>
  fetchAPI<SamplingPreviewResponse>(`${API_BASE_URL}/sampling/preview`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(requestBody),
  });

export const scoreSampling = (requestBody: SamplingScoreRequest) =>
  fetchAPI<SamplingScoreResponse>(`${API_BASE_URL}/sampling/score`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(requestBody),
  });

export const generateRecipe = (requestBody: GenerateRecipeRequest) =>
  fetchAPI<GenerateRecipeResponse>(`${API_BASE_URL}/recipes/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(requestBody),
  });
